lgc/builder/BuilderImpl.cpp

```diff
@@ -503,36 +505,115 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
+    auto descTy = desc->getType();
+#if LLVM_MAIN_REVISION && LLVM_MAIN_REVISION < 463892
+    // Old version of the code
+#else
+    // When the non-uniform use is in a VGPR, we can save a v_mov by not inserting the amdgcn_waterfall_readfirstlane
+    if (!useVgprForOperands)
+#endif
+    desc = CreateIntrinsic(Intrinsic::amdgcn_waterfall_readfirstlane, {descTy, descTy}, {waterfallBegin, desc}, nullptr,
+                           instName);
+    if (nonUniformInst->getType()->isVoidTy()) {
+      // The buffer/image operation we are waterfalling is a store with no return value. Use
+      // llvm.amdgcn.waterfall.last.use on the descriptor.
+#if LLVM_MAIN_REVISION && LLVM_MAIN_REVISION < 463892
+      // Old version of the code
+      desc = CreateIntrinsic(Intrinsic::amdgcn_waterfall_last_use, descTy, {waterfallBegin, desc}, nullptr, instName);
+#else
+      desc = CreateIntrinsic(useVgprForOperands ? Intrinsic::amdgcn_waterfall_last_use_vgpr
+                                                : Intrinsic::amdgcn_waterfall_last_use,
+                             descTy, {waterfallBegin, desc}, nullptr, instName);
+#endif
+    }
+    // Replace the descriptor operand in the buffer/image operation.
+    nonUniformInst->setOperand(operandIdx, desc);
   }
+  return waterfallBegin;
+}
 
-  return true;
+Value *BuilderImpl::emitWaterfallBeginForScalarizedLoops(Instruction *nonUniformInst, Instruction *firstIndexInst,
+                                                         ArrayRef<unsigned> operandIdxs,
+                                                         ArrayRef<Value *> nonUniformIndices, bool useVgprForOperands,
+                                                         const Twine &instName) {
+  assert(firstIndexInst);
+  auto descTy = firstIndexInst->getType();
+  SetInsertPoint(nonUniformInst);
+  Instruction *waterfallBegin = CreateIntrinsic(Intrinsic::amdgcn_waterfall_begin, descTy,
+                                                {ConstantInt::get(getInt32Ty(), 0), firstIndexInst}, nullptr, instName);
+  Instruction *readFirstLane = CreateIntrinsic(Intrinsic::amdgcn_waterfall_readfirstlane, {descTy, descTy},
+                                               {waterfallBegin, firstIndexInst}, nullptr, instName);
+  // Find the loads that should be pushed in the loop and copy them inside the waterfall loop.
+  for (unsigned operandIdx : operandIdxs) {
+    Value *desc = nonUniformInst->getOperand(operandIdx);
+    Value *OrigLoad = nullptr;
+    Value *OrigAnd = nullptr;
+    Value *OrigGep = nullptr;
+    Value *OrigExtract = nullptr;
+    if (isa<LoadInst>(desc) && match(desc, m_Load(m_Value(OrigGep)))) {
+      Instruction *newLoad = cast<LoadInst>(desc)->clone();
+      auto *newGep = dyn_cast<GetElementPtrInst>(OrigGep)->clone();
+      newLoad->insertBefore(nonUniformInst);
+      nonUniformInst->setOperand(operandIdx, newLoad);
+      newGep->insertBefore(newLoad);
+      newGep->setOperand(1, readFirstLane);
```

Comment:
Need to check that operand 1 is actually the firstIndexInst.

Reply:
I think we should not do it in the new version of the code.